src/s2/util/math/exactfloat/bignum.cc

```diff
@@ -0,0 +1,757 @@
+static Bigit SubBorrow(Bigit a, Bigit b, Bigit& borrow) {
+  Bigit diff = a - b - borrow;
+  borrow = (a < b) || (borrow && (a == b));
+  return diff;
+}
+
+// Computes a * b + c and updates the carry.
+static Bigit MulCarry(Bigit a, Bigit b, Bigit& c) {
+  auto sum = absl::uint128(a) * b + c;
+  c = absl::Uint128High64(sum);
+  return static_cast<Bigit>(sum);
+}
+
+// Computes out += a * b + c and updates the carry.
+static void MulAddCarry(Bigit& out, Bigit a, Bigit b, Bigit& c) {
+  auto sum = absl::uint128(a) * b + c + out;
+  c = absl::Uint128High64(sum);
+  out = static_cast<Bigit>(sum);
+}
+
+// Computes a += b in place. Returns the final carry (if any).
+// NOTE: the a operand must be pre-expanded to fit b.
+static Bigit AddInPlace(absl::Span<Bigit> a, absl::Span<const Bigit> b) {
+  ABSL_DCHECK_GE(a.size(), b.size());
+
+  Bigit* pa = a.data();
+  const Bigit* pb = b.data();
+
+  int left = b.size();
+  Bigit carry = 0;
+
+  // Dispatch four at a time to help loop unrolling.
+  while (left >= 4) {
+    for (int i = 0; i < 4; ++i) {
+      *pa = AddCarry(*pa, *pb++, carry);
+      ++pa;
+      --left;
+    }
+  }
+
+  // Finish remainder.
+  while (left--) {
+    *pa = AddCarry(*pa, *pb++, carry);
+    ++pa;
+  }
+
+  // Propagate carry through the rest of a.
+  int remaining = a.size() - b.size();
+  while (carry && remaining--) {
+    *pa = AddCarry(*pa, 0, carry);
+    ++pa;
+  }
+
+  return carry;
+}
+
+static ssize_t AddInto(  //
+    absl::Span<Bigit> dst, absl::Span<const Bigit> a,
+    absl::Span<const Bigit> b) {
+  const size_t max_size = std::max(a.size(), b.size());
+  const size_t min_size = std::min(a.size(), b.size());
+  ABSL_DCHECK_GE(dst.size(), max_size + 1);
+
+  Bigit* pdst = dst.data();
+  const Bigit* pa = a.data();
+  const Bigit* pb = b.data();
+
+  // Add common parts.
+  Bigit carry = 0;
+
+  // Dispatch four at a time to help loop unrolling.
+  int size = min_size;
+  int i = 0;
+  while (size >= i + 4) {
+    for (int j = 0; j < 4; ++j) {
+      pdst[i] = AddCarry(pa[i], pb[i], carry);
+      ++i;
+    }
+  }
+
+  // Finish remainder of common parts.
+  for (; i < size; ++i) {
+    pdst[i] = AddCarry(pa[i], pb[i], carry);
+  }
+
+  // Copy remaining digits from the longer operand and propagate carry.
+  auto longer = (a.size() > b.size()) ? a : b;
+  const Bigit* plonger = (a.size() > b.size()) ? pa : pb;
+
+  // Dispatch four at a time for the remaining part.
+  size = longer.size();
+  while (size >= i + 4) {
+    for (int j = 0; j < 4; ++j) {
+      pdst[i] = AddCarry(plonger[i], 0, carry);
+      ++i;
+    }
+  }
+
+  // Finish remainder.
+  for (; i < size; ++i) {
+    pdst[i] = AddCarry(plonger[i], 0, carry);
+  }
+
+  if (carry) {
+    pdst[i++] = carry;
+    return max_size + 1;
+  }
+  return max_size;
+}
+
+// Computes a -= b. Returns the final borrow (if any).
+//
+// REQUIRES: |a| < |b|.
+// NOTE: A must be pre-expanded to match the size of b.
+static Bigit SubLtIp(  //
```

Comment:
"Ip" for "InPlace" is a bit obscure.

https://google.github.io/styleguide/cppguide.html#General_Naming_Rules